Source/WebKit/UIProcess/BrowsingContextGroup.cpp

```diff
@@ -102,14 +118,35 @@ void BrowsingContextGroup::addFrameProcessAndInjectPageContextIf(FrameProcess& p
         }
 #endif
         set.add(WTFMove(newRemotePage));
+    };
+
+    if (process.isSlushBucket()) {
+        ASSERT(m_slushBucketSites.size() == 1);
+        Ref processProxy = process.process();
+        for (Ref page : m_pages)
+            createRemotePageIfNeeded(page, *m_slushBucketSites.begin());
+        return;
+    }
+    auto& site = *process.site();
+    ASSERT(!m_processMap.get(site) || m_processMap.get(site)->process().state() == WebProcessProxy::State::Terminated);
+    m_processMap.set(site, process);
+    for (Ref page : m_pages) {
+        if (site == Site(URL(page->currentURL())))
+            continue;
+        createRemotePageIfNeeded(page, site);
     }
 }
 
 void BrowsingContextGroup::removeFrameProcess(FrameProcess& process)
 {
-    ASSERT(process.site().isEmpty() || m_processMap.get(process.site()).get() == &process || process.process().state() == WebProcessProxy::State::Terminated);
-    m_processMap.remove(process.site());
-
+    if (process.isSlushBucket()) {
+        m_slushBucketProcess = nullptr;
+        m_slushBucketSites.clear();
```

Comment:
I feel like we may want to have a mechanism for removing the sites one by one instead of having a growing collection that only clears when it is completely done.  This is fine for this PR, though.

Reply:
Maybe? I don't think we can remove a site from slush bucket once it's added to it though unless all frames that had loaded the site has either been removed or navigated away.